File: java_solution/204. Count Primes.py
class Solution:
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """
        
        if n < 3:
            return 0
        
        res = [True] * n
        
        res[0] = res[1] = False
        
        for i in range(2, int(math.sqrt(n))+1):
            res[i*i:n:i] = [False] * len(res[i*i:n:i])
        
        return sum(res)
        
 
 # 统计比n小的质数有多少
 # 设置一个长度为n的list，值全部为True
 # 遍历所有非质数，标记为False
 # 统计True的个数

File: java_solution/204. Count Primes.java
// 亚马逊,微软
// 计算小于非负数n的质数的个数
// 一种比较好的质数筛选方法：埃拉托斯特尼筛法,然后定义一个boolean数组记录

class Solution {
    public int countPrimes(int n) {
        
        if (n <= 1) {
            return 0;
        }
        
        int result = 0;
        boolean[] notPrime = new boolean[n];
        
        for (int i = 2; i < n; i++) {
            if (notPrime[i] == false) {
                result++;
                
                for (int j = 2; i * j < n; j++) {
                    notPrime[i * j] = true;
                }
                
            }
        }
        return result;
    }
}
